### PR TITLE
mvcc: return -1 for wrong watcher range key >= end

### LINKED ISSUE #6819: panic: watcher not victim but not in watch groups
Steps to reproduce:

1. start a single node etcd
`vimal (master) bin $ ./etcd --debug=true`

2. start watch, and then press Ctrl-C
```
vimal (master) bin $ ./etcdctl watch key key --rev=1  -w=simple
^C
vimal (master) bin $ 
```

results in below panic :

```
panic: watcher not victim but not in watch groups

goroutine 172 [running]:
panic(0xc8b540, 0xc42000fac0)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/mvcc.(*watchableStore).cancelWatcher(0xc4201f83f0, 0xc4201dc620)
	/home/vimal/gowork/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/mvcc/watchable_store.go:236 +0x231
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/mvcc.(*watchableStore).watch.func1()
	/home/vimal/gowork/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/mvcc/watchable_store.go:218 +0x33
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/mvcc.(*watchStream).Close(0xc4203af1c0)
	/home/vimal/gowork/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/mvcc/watcher.go:143 +0xd6
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdserver/api/v3rpc.(*serverWatchStream).close(0xc420088240)
	/home/vimal/gowork/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdserver/api/v3rpc/watch.go:346 +0x34
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdserver/api/v3rpc.(*watchServer).Watch(0xc4203ee240, 0x128c000, 0xc42000f5b0, 0xc420571b28, 0xc420571b38)
	/home/vimal/gowork/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdserver/api/v3rpc/watch.go:145 +0x3a2
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdserver/etcdserverpb._Watch_Watch_Handler(0xcc8bc0, 0xc4203ee240, 0x128a380, 0xc4200881b0, 0x128c0c0, 0xc4203af180)
	/home/vimal/gowork/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdserver/etcdserverpb/rpc.pb.go:2455 +0xbb
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdserver/api/v3rpc.metricsStreamInterceptor(0xcc8bc0, 0xc4203ee240, 0x128a380, 0xc4200881b0, 0xc420122600, 0xe830d0, 0xc42002f800, 0x0)
	/home/vimal/gowork/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdserver/api/v3rpc/interceptor.go:113 +0x12c
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdserver/api/v3rpc.newStreamInterceptor.func1(0xcc8bc0, 0xc4203ee240, 0x128a380, 0xc4200881b0, 0xc420122600, 0xe830d0, 0x0, 0x0)
	/home/vimal/gowork/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdserver/api/v3rpc/interceptor.go:91 +0x115
github.com/coreos/etcd/cmd/vendor/google.golang.org/grpc.(*Server).processStreamingRPC(0xc420460000, 0x128c240, 0xc420088120, 0xc4203c43c0, 0xc4203ee2d0, 0x12a5420, 0x0, 0x0, 0x0)
	/home/vimal/gowork/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/google.golang.org/grpc/server.go:693 +0x2b6
github.com/coreos/etcd/cmd/vendor/google.golang.org/grpc.(*Server).handleStream(0xc420460000, 0x128c240, 0xc420088120, 0xc4203c43c0, 0x0)
	/home/vimal/gowork/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/google.golang.org/grpc/server.go:770 +0xc36
github.com/coreos/etcd/cmd/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc42000f460, 0xc420460000, 0x128c240, 0xc420088120, 0xc4203c43c0)
	/home/vimal/gowork/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/google.golang.org/grpc/server.go:419 +0xab
created by github.com/coreos/etcd/cmd/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
	/home/vimal/gowork/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/cmd/vendor/google.golang.org/grpc/server.go:420 +0xa3


```

Environment : 
Ubuntu 14.04 LTE, with HDD

### PR DESCRIPTION
**Pull request by gyuho**
Fix #6819.

@heyitsanthony @xiang90 Should we translate the error in etcdctl or clientv3 side? Currently clients will just see the watch create request is canceled.

### CODE REVIEW

File: mvcc/watcher.go [hunk 1]
```diff
@@ -108,6 +109,10 @@ func (ws *watchStream) Watch(key, end []byte, startRev int64, fcs ...FilterFunc)
 	id := ws.nextID
 	ws.nextID++
 
+	// prevent wrong range where key >= end lexicographically
```
Comment (heyitsanthony) [inline mvcc/watcher.go]: check before acquiring the lock?

**Comment by heyitsanthony** [comment]
No strong opinion about it being translated in clientv3/etcdctl. We should probably teach etcdctl about watchresponse Err() regardless.

**Comment by gyuho** [comment]
@heyitsanthony @xiang90 

All fixed. Think error `Error:  watch is canceled by the server` is good enough?
I instead added more clarification on range end docs for clientv3 and etcdctl.

PTAL.

Thanks.

File: etcdctl/ctlv3/command/watch_command.go [hunk 2]
```diff
@@ -36,7 +36,7 @@ var (
 func NewWatchCommand() *cobra.Command {
 	cmd := &cobra.Command{
 		Use:   "watch [options] [key or prefix] [range_end]",
-		Short: "Watches events stream on keys or prefixes",
+		Short: "Watches events stream on keys or prefixes (range_end is optional or must be lexicographically greater than key)",
```
Comment (heyitsanthony) [inline etcdctl/ctlv3/command/watch_command.go]: this is sort of wordy, probably can drop

File: mvcc/watcher.go [hunk 3]
```diff
@@ -99,6 +100,11 @@ type watchStream struct {
 // Watch creates a new watcher in the stream and returns its WatchID.
 // TODO: return error if ws is closed?
 func (ws *watchStream) Watch(key, end []byte, startRev int64, fcs ...FilterFunc) WatchID {
+	// prevent wrong range where key >= end lexicographically
+	if len(end) != 0 && bytes.Compare(key, end) != -1 {
```
Comment (heyitsanthony) [inline mvcc/watcher.go]: does this break `Watch(ctx, "abc", clientv3.WithFromKey())`?

**Comment by gyuho** [comment]
@heyitsanthony Added special case handling for `[]byte{0}` with test. PTAL. Thanks!

File: mvcc/watcher.go [hunk 4]
```diff
@@ -99,6 +100,12 @@ type watchStream struct {
 // Watch creates a new watcher in the stream and returns its WatchID.
 // TODO: return error if ws is closed?
 func (ws *watchStream) Watch(key, end []byte, startRev int64, fcs ...FilterFunc) WatchID {
+	// prevent wrong range where key >= end lexicographically
+	// except special case for watch from key '\x00'
+	if len(end) != 0 && !bytes.Equal(end, []byte{0}) && bytes.Compare(key, end) != -1 {
```
Comment (heyitsanthony) [inline mvcc/watcher.go]: `if len(end) != 0 && bytes.Compare(key, end) != -1 &&  !bytes.Equal(end, []byte{0})`? common path is going to be key < end, right?

File: mvcc/watcher_test.go [hunk 5]
```diff
@@ -153,6 +153,28 @@ func TestWatcherWatchPrefix(t *testing.T) {
...
+	defer w.Close()
+
+	if id := w.Watch([]byte("foa"), []byte("foa"), 1); id != -1 {
+		t.Fatalf("key == end range given; id expected -1, got %d", id)
+	}
+	if id := w.Watch([]byte("fob"), []byte("foa"), 1); id != -1 {
+		t.Fatalf("key > end range given; id expected -1, got %d", id)
+	}
+	// in case client sends 'WithFromKey' request with smallest bytes range end
+	if id := w.Watch([]byte("foo"), []byte{}, 1); id != 0 {
```
Comment (heyitsanthony) [inline mvcc/watcher_test.go]: should this be []byte{0}? I think []byte{} will give the same result as end == nil once passed through grpc?

**Comment by gyuho** [comment]
@heyitsanthony I removed `bytes.Equal` check because we convert `\x00` into empty bytes in v3rpc https://github.com/coreos/etcd/blob/master/etcdserver/api/v3rpc/watch.go#L173. What do you think?

It's just empty bytes when passed to mvcc.

**Comment by heyitsanthony** [comment]
@gyuho OK. Is that WithFromKey correct, though? I think it still needs to be `[]byte{0}` so the v3rpc/watch stuff can convert it to []byte{}?

**Comment by gyuho** [comment]
@heyitsanthony Did you mean `TestWatcherWatchWrongRange` to use `w.Watch([]byte("foo"), []byte{0}, 1)`?  How so, we don't use rpc in mvcc tests?

I added `[]byte{0}` case to integration test to make sure we don't break `WithFromKey`

**Comment by heyitsanthony** [comment]
Oh oops, you're right. What was I thinking. lgtm

**Comment by gyuho** [comment]
Thanks merging in greens.

**Pull request merged**